packages/ra-ui-materialui/src/layout/Notification.tsx

```diff
@@ -43,6 +43,14 @@ export const Notification = (props: NotificationProps) => {
     const translate = useTranslate();
 
     useEffect(() => {
+        const beforeunload = (e: BeforeUnloadEvent) => {
```

Comment:
This will always prevent the default, no matter if there is an undoable operation ongoing or not.

Reply:
I just did a test and it seems to work thanks to the condition below

Reply:
I suggest you only add an event listener when you're in an undoable notification, instead of always adding it and verifying the condition in the listener. Apparently, beforeunload event listeners affect performance (see https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#usage_notes) and should be used with caution. 

Also, the advised syntax for a beforeunload event listener is to return a string (see https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#examples). Some browsers don't follow the spec, and the preventDefault won't be enough.

Reply:
You are right, add the event listener in the condition is working as expected. I may have made a careless mistake.
I also updated the return type to maximize browsers compatibility according to [this examples](https://developer.mozilla.org/en-US/docs/Web/API/BeforeUnloadEvent#examples)